src/v/cluster/plugin_frontend.h

```diff
@@ -0,0 +1,161 @@
+    ss::future<mutation_result>
+      remove_transform(model::transform_name, model::timeout_clock::time_point);
+
+    // Register for updates going forward.
+    //
+    // This is a core local operation, you can register for updates to the
+    // transforms table. If you want to lookup the transform you can then do
+    // that in the callback.
+    notification_id register_for_updates(notification_callback);
+
+    // Unregister a previously registered notification.
+    void unregister_for_updates(notification_id);
+
+    // Lookup a transform by ID. Useful when paired with registering for
+    // updates.
+    std::optional<model::transform_metadata>
+      lookup_transform(model::transform_id) const;
+
+    // Lookup transforms for input topics.
+    absl::btree_map<model::transform_id, model::transform_metadata>
+      lookup_transforms_by_input_topic(model::topic_namespace_view) const;
+
+    // Get a snapshot of all the transforms that currently exist.
+    absl::btree_map<model::transform_id, model::transform_metadata>
+    all_transforms() const;
+
+private:
+    // Perform a mutation request, check if this node is the cluster leader and
+    // ensuring that the right core is used for local mutations.
+    ss::future<mutation_result>
+      do_mutation(transform_cmd, model::timeout_clock::time_point);
+
+    // If this node is not the leader, dispatch this mutation to `node_id` as
+    // the cluster's leader.
+    ss::future<mutation_result> dispatch_mutation_to_remote(
+      model::node_id, transform_cmd, model::timeout_clock::duration);
+
+    // Performs (and validates) a mutation command to be inserted into the
+    // controller log.
+    //
+    // This must take place on the controller shard on the cluster leader.
+    ss::future<mutation_result>
+      do_local_mutation(transform_cmd, model::timeout_clock::time_point);
+
+    // Ensures that the mutation is valid.
+    //
+    // This must take place on the controller shard on the cluster leader while
+    // holding _mu to ensure consistency.
+    errc validate_mutation(const transform_cmd&);
+
+public:
+    /**
+     * A validator of commands that come through, the validation must happen on
+     * the controller leader, core 0 so that these is a consistent view of the
+     * data in the plugin table.
+     *
+     * Mostly exposed seperately for unit testing.
+     */
+    class validator {
+    public:
+        validator(topic_table*, plugin_table*);
+
+        // Ensures that the mutation is valid.
+        //
+        // NOTE: At the time of writing the topic validation is best effort,
+        // it is possible to race with mutations in the topics_frontend and
+        // topics can be deleted or have partitions added that wouldn't be
+        // valid otherwise.
+        errc validate_mutation(const transform_cmd&);
+
+    private:
+        // Would adding this input->output transform cause a cycle?
+        bool would_cause_cycle(
```

Comment:
i'm guessing there isn't any risk of out-of-band operations (e.g. create topic) that would invalidate this check after it returns success?

Reply:
Not this check no. All mutations to plugin_frontend are done under a mutex and we're only checking this data, so I believe this invariant can't be broken, only the two that I posted as a comment that @bharathv and I discussed.